posthog/hogql_queries/actors_query_runner.py

```diff
@@ -230,21 +239,67 @@ def to_query(self) -> ast.SelectQuery:
                 order_by = []
 
         with self.timings.measure("select"):
-            if self.query.source:
-                join_expr = self.source_table_join()
-            else:
+            if not self.query.source:
                 join_expr = ast.JoinExpr(table=ast.Field(chain=[self.strategy.origin]))
+            else:
+                assert self.source_query_runner is not None  # For type checking
+                source_query = self.source_query_runner.to_actors_query()
 
-            stmt = ast.SelectQuery(
-                select=columns,
-                select_from=join_expr,
-                where=where,
-                having=having,
-                group_by=group_by if has_any_aggregation else None,
-                order_by=order_by,
-            )
+                source_id_chain = self.source_id_column(source_query)
+                source_alias = "source"
+
+                origin = self.strategy.origin
 
-        return stmt
+                join_on: ast.Expr = ast.CompareOperation(
+                    op=ast.CompareOperationOp.Eq,
+                    left=ast.Field(chain=[origin, self.strategy.origin_id]),
+                    right=ast.Field(chain=[source_alias, *source_id_chain]),
+                )
+
+                if isinstance(self.strategy, PersonStrategy):
+                    # SelectUnionQuery (used by Stickiness) doesn't have settings
+                    for select_query in (
+                        [source_query] if isinstance(source_query, SelectQuery) else source_query.select_queries
+                    ):
+                        if select_query.settings is None:
+                            select_query.settings = HogQLQuerySettings()
+                        select_query.settings.use_query_cache = True
+                        select_query.settings.query_cache_ttl = HOGQL_INCREASED_MAX_EXECUTION_TIME
+                    join_on = ast.And(
+                        exprs=[
+                            join_on,
+                            ast.CompareOperation(
+                                left=ast.Field(chain=[origin, "id"]),
+                                right=ast.SelectQuery(
+                                    select=[ast.Field(chain=[source_alias, *self.source_id_column(source_query)])],
+                                    select_from=ast.JoinExpr(table=source_query, alias=source_alias),
+                                ),
+                                op=ast.CompareOperationOp.In,
+                            ),
+                        ]
+                    )
```

Comment:
This part could use some comments "doing x because of y" – I feel like in a few months whoever will be reading this, will need a session of `git blame` PR tracing just to figure this out, which comments can preempt